File: algorithm/002-add_two_numbers/scala/solution.scala
class ListNode(var _x: Int = 0) {
  var next: ListNode = null
  var x: Int = _x
}

object Solution {
  def addTwoNumbers(l1: ListNode, l2: ListNode): ListNode = {
    var sum = new ListNode(0)
    val head = sum
    var carry = 0

    var p = l1
    var q = l2
    while (p != null || q != null || carry > 0) {
      if (p != null) {
        sum.x += p.x
        p = p.next
      }

      if (q != null) {
        sum.x += q.x
        q = q.next
      }

      sum.x += carry
      carry = sum.x / 10
      sum.x %= 10

      if (p != null || q != null || carry > 0) {
        sum.next = new ListNode(0)
        sum = sum.next
      }
    }

    return head
  }
}

File: algorithm/002-add_two_numbers/swift/solution.swift
public class ListNode {
    public var val: Int
    public var next: ListNode?
    public init(_ val: Int) {
        self.val = val
        self.next = nil
    }
}

class Solution {
    func addTwoNumbers(_ l1: ListNode?, _ l2: ListNode?) -> ListNode? {
        var sum = ListNode(0)
        let head = sum
        var carry = 0
        
        var l1 = l1
        var l2 = l2
        while (l1 != nil) || (l2 != nil) || carry > 0 {
            if l1 != nil {
                sum.val += l1!.val
                l1 = l1!.next
            }
            
            if l2 != nil {
                sum.val += l2!.val
                l2 = l2!.next
            }
            
            sum.val += carry
            carry = sum.val / 10
            sum.val %= 10
            
            if l1 != nil || l2 != nil || carry > 0 {
                sum.next = ListNode(0)
                sum = sum.next!
            }
        }
        
        return head
    }
}